zoo/birealnet.py

```diff
@@ -0,0 +1,212 @@
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__author__ = "Francesco Loro"
+__email__ = "francesco.official@gmail.com"
+__supervisor__ = "Danilo Pau"
+__email__ = "danilo.pau@st.com"
+
+# Download pretrained weight from:
+# Birealnet -> https://drive.google.com/file/d/1BuDYhydNAy-sFdvoh24gJbpA0zRDvGVq/view?usp=sharing
+
+import qkeras as q
+import tensorflow as tf
+import larq as lq
+from utils import compare_network, create_random_dataset, dump_network_to_json
+
+# Define path to the pre-trained weights
+PATH_BIREALNET = "weights/birealnet_weights.h5"
+BIREALNET_NAME = "biRealNet"
+
+
+class BirealNet:
+  """
+  Class to create and load weights of: biRealNet
+  Attributes:
+        network_name: Name of the network
+  """
+
+  def __init__(self):
+    self.__weights_path = PATH_BIREALNET
+    self.network_name = BIREALNET_NAME
+
+  @staticmethod
+  def add_qkeras_residual_block(model, filters_num, strides=1):
+    """
+    Add a sequence of: Activation quantization, Quantized Conv2D
+    :param model: model where to add the sequence
+    :param filters_num: number of filters for Cov2D
+    :param strides: strides for Conv2D
+    """
+    model.add(q.QActivation("binary(alpha=1)"))
+    model.add(q.QConv2D(filters_num, (3, 3), strides=strides, padding="same",
+                        kernel_quantizer="binary(alpha=1)", use_bias=False))
+    model.add(tf.keras.layers.BatchNormalization())
+
+  @staticmethod
+  def add_qkeras_connection_block(model, filters_num):
+    """
+    Add a sequence of: Activation quantization, Quantized Conv2D, reshape,
+    Average Pooling, Conv2D, 2x BatchNormalization
+    :param model: model where to add the sequence
+    :param filters_num: number of filters for Cov2D
+    """
+    model.add(q.QActivation("binary"))
+    model.add(q.QConv2D(filters_num, (3, 3), strides=(2, 2), use_bias=False,
+                        padding="same", kernel_quantizer="binary(alpha=1)"))
+    # Prepare shapes for reshape layers
+    shape_in = model.output_shape[1] * model.output_shape[2] * \
+               model.output_shape[3]
+    shape_out = (model.output_shape[1], model.output_shape[2],
+                 model.output_shape[3] // 2)
+    model.add(tf.keras.layers.Flatten())
+    model.add(tf.keras.layers.Reshape(target_shape=(shape_in, 1)))
+    model.add(tf.keras.layers.AvgPool1D(1, strides=2, padding="same"))
+    model.add(tf.keras.layers.Reshape(target_shape=shape_out))
+    model.add(tf.keras.layers.Conv2D(filters_num, (1, 1), padding="same",
+                                     use_bias=False))
+    model.add(tf.keras.layers.BatchNormalization())
+    model.add(tf.keras.layers.BatchNormalization())
```

Comment:
why consecutive batchnormalization layers?

Reply:
In next commit I've done a complete refactor of that class, fixing also that mistake